cpp/oneapi/dal/algo/basic_statistics/backend/gpu/compute_kernel_dense_impl_dpc.cpp

```diff
@@ -0,0 +1,1230 @@
+                const std::int64_t gid = item.get_group().get_id(0);
+
+                const std::int64_t row_block_idx = gid / column_block_count;
+                const std::int64_t col_block_idx = gid - row_block_idx * column_block_count;
+
+                block_processor<Float, List>(data_ptr,
+                                             ablock_rc_ptr,
+                                             amin_ptr,
+                                             amax_ptr,
+                                             asum_ptr,
+                                             asum2_ptr,
+                                             asum2cent_ptr,
+                                             row_count,
+                                             row_block_idx,
+                                             row_block_count,
+                                             column_count,
+                                             col_block_idx,
+                                             column_block_count,
+                                             tid,
+                                             tnum);
+            });
+        });
+    }
+
+    auto [ndres, merge_event] =
+        merge_blocks(std::move(ndbuf), column_count, row_block_count, { last_event });
+
+    return std::make_tuple(std::move(ndres), merge_event);
+}
+
+template <typename Float, bs_list List>
+std::tuple<ndresult<Float, List>, sycl::event> compute_kernel_dense_impl<Float, List>::finalize(
+    ndresult_t&& ndres,
+    std::int64_t row_count,
+    std::int64_t column_count,
+    const bk::event_vector& deps) {
+    ONEDAL_PROFILER_TASK(finalize, q_);
+
+    const bool distr_mode = comm_.get_rank_count() > 1;
+    // ndres asserts
+    ASSERT_IF(bs_list::min, ndres.get_min().get_count() == column_count);
+    ASSERT_IF(bs_list::max, ndres.get_max().get_count() == column_count);
+    if (distr_mode) {
+        ASSERT_IF(bs_list::mean | sum2cent_based_stat, ndres.get_sum().get_count() == column_count);
+    }
+    else {
+        ASSERT_IF(bs_list::sum, ndres.get_sum().get_count() == column_count);
+    }
+    ASSERT_IF(bs_list::sum2 | bs_list::sorm, ndres.get_sum2().get_count() == column_count);
+    if (distr_mode) {
+        ASSERT_IF(bs_list::varc | bs_list::stdev | bs_list::vart,
+                  ndres.get_sum2cent().get_count() == column_count);
+    }
+    else {
+        ASSERT_IF(bs_list::sum2cent, ndres.get_sum2cent().get_count() == column_count);
+    }
+
+    sycl::event last_event;
+    if (distr_mode) {
+        if constexpr (check_mask_flag(bs_list::min, List)) {
+            comm_
+                .allreduce(q_,
+                           ndres.get_min().get_data(),
+                           ndres.get_min().get_mutable_data(),
+                           ndres.get_min().get_count(),
+                           de::v1::spmd_reduce_op::min,
+                           deps)
+                .wait();
```

Comment:
Can it be done in async mode?

Reply:
It is necessary to think about infrastructure for that. Right now there is no something like sycl::event::wait(deps)

Reply:
Agree, the infrastructure is required, will migrate to asynk for the next optimization iteration when required helpers is implemented